e2e-testing.md

```diff
@@ -0,0 +1,321 @@
+- [3\. Test scenarios](#2-test-scenarios)
+  - [3.1. Use snake_case for test folders naming](#31-use-snake-case-for-test-folders-naming)
+  - [3.2. Use folders for test suite organization](#32-use-folders-for-test-suite-organization)
+  - [3.3. Use one spec file per one test](#33-use-one-spec-per-one-test)
+  - [3.4. Before block should not contain common test data preparation](#34-before-block-should-not-contain-common-test-data-preparation)
+  - [3.5. Test block should only contain test steps_and_assertions](#35-test-block-should-only-contain-test-steps-and-assertions)
+- [4\. Assertions](#2-assertions)
+  - [4.1 One assertion per one test step method](#41-one-assertion-per-one-test-step-method)
+  - [4.2 Use expect for assertions](#42-use-expect-for-assertions)
+
+
+1\. Factory classes
+--------------
+
+#### 1.1. Use Factory classes only for Entities
+
+The ```FactoryItem``` class should be extended for each new Entity like User, Course, Event, StudentGroup class.
+For the tables which contains additional information for the Item it's recommended to create a ```link``` method.
+
+
+```typescript
+// ✅ good
+export class User extends FactoryItem {
+
+  public constructor(
+    apiClient: SeedAPI,
+    options: Options,
+  ) {
+    super(apiClient);
+    ...
+  }
+
+  async linkToCourse(
+    courseId: number,
+    status: CourseUserStatus,
+  ): Promise<void> {
+    await this.api.seedItemIfNotExist(
+      FactoryItemType.CourseUsers,
+      {
+        courseId,
+        userId: this.id,
+        status,
+      },
+    );
+  }
+}
+```
+
+2\. Page objects
+--------------
+
+#### 2.1. Define page locators outside the constructor
+
+Please define locators in the page object classes and components using the below example.
+There could be different approaches, but we agreed to follow a single style within the organization.
+
+```typescript
+// ❌ bad
+export class SignInPage extends BasePage{
+  private readonly emailField: Locator;
+
+  constructor(page: Page) {
+      super(page);
+      this.emailField = page.getByTestId('sign-in-user-email');
+  }
+}
+
+// ✅ good
+export class SignInPage extends BasePage{
+  private readonly emailField = this.page.getByTestId('sign-in-user-email');
+
+  constructor(page: Page) {
+    super(page);
+  }
+}
+```
+
+
+
+#### 2.2. Parametrized page URLs defining
+
+Use the below approach to define the parametrized page object URLs.
+
+```typescript
+// ✅ good
+interface Options {
+  chatId: number;
+}
+
+export class ChatPage extends LoggedInBasePage {
+
+  constructor(
+    page: Page,
+    options: Options,
+  ) {
+    super(page);
+
+    this.url = ROUTES.chat(options.chatId).index;
+  }
+}
+```
+
+
+#### 2.3. Use Components to re-use common page element groups
+
+
+If you defined some common elements from the header, footer, sideBar, or some popups that appear on several pages - make sure to define them in the corresponding component class.
```

Comment:
it's more about DRY pattern, right?

ideally that all UI components (modals, etc) have their representations in e2e tests
it will perfectly match the design system

Reply:
Yes, that's about DRY :)